server/src-lib/Hasura/RQL/DML/Mutation.hs

```diff
@@ -90,8 +98,9 @@ mutateAndFetchCols qt cols (cte, p) strfyNum =
 
 mkSelCTEFromColVals
   :: (MonadError QErr m)
-  => QualifiedTable -> [PGColumnInfo] -> [ColVals] -> m S.CTE
-mkSelCTEFromColVals qt allCols colVals =
+  => (PGColumnType -> Value -> m (WithScalarType PGScalarValue))
+  -> QualifiedTable -> [PGColumnInfo] -> [ColVals] -> m S.CTE
+mkSelCTEFromColVals parseFn qt allCols colVals =
```

Comment:
Just for my own understanding: could you explain to me from a high-level what this function is doing? When I was first looking through this diff, I was slightly confused, because it seems like we’re parsing the values (using `parseFn`) but then immediately re-encoding them (using `toTxtValue`). It seems like `mutateAndSel` is doing that as part of some kind of two-pass query strategy, but it’s not quite clicking for me what the reason for doing that is.

Reply:
To be honest, this diff is not so informative. I'll give you a bit of context. 

`toTxtValue` will encode any `PGScalarValue` to literal values. I thought it is enough and went for #3198. In case of json/jsonb values, the json string we're getting, let's say some `String t` where `t` is json string, is parsed into `PGScalarValue` as `PGValJSON (String t)`.  The `toTxtValue` converts `PGValJSON (String t)` to json encoding of a string value. The `{"age": 23}` is converted to something like `"{\"age\": 23}"`. In `mutateAndSel` function, I'm enforcing all `String t` to `PGValText t`, hence `toTxtValue` encodes correctly for json/jsonb values. 

I hope you understand the above explanation. I'm happy to know any improvements to the current fix.

Reply:
That definitely *is* helpful, so thank you for that explanation, but it actually doesn’t answer my original question (which is only semi-related to this pull request). Specifically, I’m trying to understand what `mkSelCTEFromColVals` is doing in the first place, from a high level. 🙂 Sorry if I didn’t make that very clear!

Reply:
Okay :slightly_smiling_face: . The `mkSelCTEFromColVals` will generate a `SELECT` query, that is used inside a `WITH ....` statement. The `SELECT` statement look like
```sql
SELECT * FROM VALUES ((col1-val, col2-val, col3-val), .....) AS "table_name" (col1, col2, col3)
```
The `parseFn` converts JSON `value` to `col-val`.

Reply:
I probably should have been more clear: I can mostly make out *what* it does, but what I’m less sure about is *why*. If `mutateAndFetchCols` does the actual mutation query, and it returns a bunch of values, why don’t we parse them directly from that result? Why do we send them back through Postgres another time?

Reply:
>  Why do we send them back through Postgres another time?

To join those values and fetch associated relationships in `returning`. Note that, this code path is executed only when relationships are selected in `returning` (See `hasNestedFld $ _mFields mut`) 

I'm happy to refactor this part of code to make it much clear and self-descriptive. :)

Reply:
Ahh, I see—the reason we have to do an extra `SELECT` is the information can be in related tables that aren’t actually part of the `INSERT`/`UPDATE`. Is that right?

In that case, it seems like we either (1) need to make sure our parsing logic for SQL values is round-trippable, or (2) we should avoid parsing the values we get from Postgres the first time and just store them in the textual format Postgres already knows about. Is that feasible? The approach in this PR seems a little indirect to me.

Reply:
The (1) is possible only if we have the context of array types which #2243 is meant to solve. I think this PR is solving (2). We're fetching column values as Strings (except for geometry/geography types) and storing them as `PGValText`. The `parseJSONStringAsLiteral` is doing the job.

Reply:
> I think this PR is solving (2). We're fetching column values as Strings (except for geometry/geography types) and storing them as `PGValText`. The `parseJSONStringAsLiteral` is doing the job.

Okay, I think that helps me to understand better what’s going on. A followup question: why are we treating the geometry/geography types differently? They have a textual representation as well, right? So it seems as though we ought to be able to treat everything uniformly.

Reply:
Yes, we can use the textual representation for geometry/geography columns. Ideally, the returned columns are `HashMap PGCol Text`. I'll make the necessary change in next commit.

Reply:
Now geometry/geography values are stored as Strings via https://github.com/hasura/graphql-engine/pull/3375/commits/0f484df408336bb9827ad5bae845d98e8ce24594. Please approve if everything is OK. :slightly_smiling_face:

Reply:
@lexi-lambda Updated the PR with the latest master. Please review. :slightly_smiling_face: